server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -1136,7 +1137,15 @@ private static void validateCompositeTemplate(final ClusterState state,
                         validateTimestampFieldMapping(tsFieldName, mapperService);
                     }
                 } catch (Exception e) {
-                    throw new IllegalArgumentException("invalid composite mappings for [" + templateName + "]", e);
+                    final String expectedMessageIfDataStreamsIsUsedAndNotSupported =
+                        "Failed to parse mapping: Root mapping definition has unsupported parameters:  [_timestamp : {path=@timestamp}]";
+                    if (expectedMessageIfDataStreamsIsUsedAndNotSupported.equals(e.getMessage())) {
+                        // Fail like a parsing expection, since we will be moving data_stream template out of server module and
+                        // then we would fail with the same error message, like we do here.
+                        throw new XContentParseException("[index_template] unknown field [@timestamp]");
```

Comment:
I think we should fail with a parsing error like `"[index_template] unknown field [data_stream]"` to make it clearer that the entire data stream configuration is unsupported, rather than just the `@timestamp` part